lib/rules/no-constructor-return.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * @fileoverview Rule to disallow returning value from constructor.
+ * @author Pig Fang <https://github.com/g-plane>
+ */
+
+"use strict";
+
+const astUtils = require("./utils/ast-utils");
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        type: "suggestion",
+
+        docs: {
+            description: "disallow returning value from constructor",
+            category: "Best Practices",
+            recommended: false,
+            url: "https://eslint.org/docs/rules/no-constructor-return"
+        },
+
+        schema: {},
+
+        fixable: null,
+
+        messages: {
+            unexpected: "Unexpected return statement in constructor."
+        }
+    },
+
+    create(context) {
+        return {
+            "MethodDefinition[kind='constructor'] ReturnStatement"(node) {
+                const ancestors = context.getAncestors().reverse();
+                const ctorIndex = ancestors.findIndex(
+                    ancestor => ancestor.type === "MethodDefinition" && ancestor.kind === "constructor"
+                );
+                const upperFnIndex = ancestors.findIndex(astUtils.isFunction);
+
+                if (~upperFnIndex && upperFnIndex < ctorIndex - 1) {
+                    return;
+                }
+
+                if (
+                    ancestors[2] &&
```

Comment:
This feels a bit magical to me (how would a future maintainer know to use `ancestors[2]`)?

Would it make more sense to use the code path analysis events to identify when we've entered a new function scope? We could maintain a stack of function scopes-- and then, on any ReturnStatement, check the top of the function scope stack to see if it is a MethodDefinition that is a constructor, and if so, report.

I'm not strongly opposed to the current approach, but would you please explore this and see if it might be a potential readability improvement?